napari/_qt/_tests/test_qt_utils.py

```diff
@@ -101,3 +104,23 @@ def test_qt_might_be_rich_text(qtbot):
     qtbot.addWidget(widget)
     assert qt_might_be_rich_text("<b>rich text</b>")
     assert not qt_might_be_rich_text("plain text")
+
+
+def test_thread_proxy_guard(monkeypatch, qapp):
+    class X:
+        a = 1
+
+    monkeypatch.setenv('NAPARI_ENSURE_PLUGIN_MAIN_THREAD', 'True')
+    single_threaded_executor = ThreadPoolExecutor(max_workers=1)
```

Comment:
Probably best to make this a fixture that calls `shutdown` on teardown (credit to @kcpevey): 

```python
@pytest.fixture()
def single_threaded_executor():
    executor = ThreadPoolExecutor(max_workers=1)
    yield executor
    executor.shutdown()
```

Or use the executor in a context when it's needed:

```python
with ThreadPoolExecutor() as executor:
    f = executor.submit(x.__setattr__, 'a', 2)
    f.result()
    assert x.a == 2

    f = executor.submit(x_proxy.__setattr__, 'a', 3)
    with pytest.raises(RuntimeError):
        f.result()
```